README.md
# TEDDY - Transducer Electronic Datasheet Decryption Yielder

## Current capabilities in Version 0.4

### Supported TEDS rom device types

* DS2430A (0x14)
* DS2431 (0x2D)

### Supported TEDS versions & templates

* v.0.9 Legacy
  * Accelerometer (Template ID 0)

### Tested devices
| Manufacturer | Model number | ROM device type | TEDS version | Template ID | Read serial number | Read sensor direction | Read sensitivity | Read transducer type | Read Manufacturer | Read calibration date |
|--------------|--------------|-----------------|--------------|-------------|--------------------|-----------------------|------------------|----------------------|-------------------|-----------------------|
| **KISTLER** | 8688A50T | DS2430A (0x14) | v.0.9 | 0 | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
| **KISTLER** | 8688A50T | DS2431 (0x2D)  | v.0.9 | 0 | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
| **KISTLER** | 8762A10T | DS2430A (0x14) | v.0.9 | 0 | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |

## Links

Convert calibration date: https://www.topster.net/calendar/tagerechner.php?styp=datum&smonat=1&stag=1&sjahr=1998&emonat=&etag=&ejahr=&typ=dazu&subDazu=%2B&dazu=0

IEEE Manufacturer IDs: https://regauth.standards.ieee.org/standards-ra-web/pub/view.html#registries


CHANGELOG.md
# Changelog for TEDDY

## Changes in Version 0.4
- Decrypts the transducer manufacturer
- Decrypts the transducer calibration date

## Changes in Version 0.3
- Decrypts the transducer type (model number + model version + model version number)
- Memory optimization

## Changes in Version 0.2
- Decrypts the sensitivity of acceleration sensors with legacy TEDS version 0.9 and template Id 0
- Refactoring to not only include acceleration sensors

## Changes in Version 0.1
- Reads TEDS from devices 0x2D and 0x14 implementing legacy version 0.9 with template Id 0
- Decrypts the serial number of connected sensors

src/TEDDY.h
/*
 TEDDY.h
 
 Transducer Electronic Datasheet Decryption Yielder - An arduino library to read and decode TEDS from acceleration sensors.
 
 */

#ifndef TEDDY_h
#define TEDDY_h

#include <Arduino.h>
#include <OneWire.h>
#include <String.h>
#include <ManufacturerIDs.h>

class TEDDY {
private:
  byte TEDSROM[8];
  byte TEDSData[32];
  int busPin;
  byte offset;
  byte ROMDeviceType;
  byte SensorDirection;
  int ModelNumber;
  char ModelVersion;
  byte ModelVersionNumber;
  String TransducerType;
  TEDSManID TEDSManIDArray[150];

public:
  TEDDY(int busPin);
  ~TEDDY();
  void POPULATETEDDYMANID();
  uint8_t read();
  char* getROMDeviceType();
  unsigned long int getSerialNumber();
  char getSensorDirection();
  float getSensitivity();
  String getTransducerType();
  String getManufacturer();
  unsigned int getCalibrationDate();
};

#endif


src/ManufacturerIDs.cpp
#include "ManufacturerIDs.h"
#include <TEDDY.h>

void TEDDY::POPULATETEDDYMANID(){

	TEDSManIDArray[0].ManID = 10799;
	TEDSManIDArray[0].ManName = "SENTHER TECHNOLOGIES";
	TEDSManIDArray[1].ManID = 1117;
	TEDSManIDArray[1].ManName = "SKC Audio";
	TEDSManIDArray[2].ManID = 1107;
	TEDSManIDArray[2].ManName = "DongHua Testing Technology Co., Ltd.";
	TEDSManIDArray[3].ManID = 1105;
	TEDSManIDArray[3].ManName = "EnvyLogic Co.,Ltd.";
	TEDSManIDArray[4].ManID = 106;
	TEDSManIDArray[4].ManName = "Dewesoft d.o.o.";
	TEDSManIDArray[5].ManID = 105;
	TEDSManIDArray[5].ManName = "IPETRONIK GmbH & Co.KG";
	TEDSManIDArray[6].ManID = 102;
	TEDSManIDArray[6].ManName = "tecsis LP";
	TEDSManIDArray[7].ManID = 93;
	TEDSManIDArray[7].ManName = "Monitran";
	TEDSManIDArray[8].ManID = 94;
	TEDSManIDArray[8].ManName = "Test GmbH";
	TEDSManIDArray[9].ManID = 66;
	TEDSManIDArray[9].ManName = "Sinocera Piezotronics, Inc.";
	TEDSManIDArray[10].ManID = 1097;
	TEDSManIDArray[10].ManName = "ETAS";
	TEDSManIDArray[11].ManID = 68;
	TEDSManIDArray[11].ManName = "VIP Sensors";
	TEDSManIDArray[12].ManID = 44;
	TEDSManIDArray[12].ManName = "Fuji Ceramics";
	TEDSManIDArray[13].ManID = 59;
	TEDSManIDArray[13].ManName = "Interface, Inc.";
	TEDSManIDArray[14].ManID = 54;
	TEDSManIDArray[14].ManName = "imc MeBsysteme GmbH";
	TEDSManIDArray[15].ManID = 34;
	TEDSManIDArray[15].ManName = "Esensors";
	TEDSManIDArray[16].ManID = 49;
	TEDSManIDArray[16].ManName = "Sensorex";
	TEDSManIDArray[17].ManID = 37;
	TEDSManIDArray[17].ManName = "Lebow";
	TEDSManIDArray[18].ManID = 42;
	TEDSManIDArray[18].ManName = "Transducer Techniques";
	TEDSManIDArray[19].ManID = 47;
	TEDSManIDArray[19].ManName = "LEM";
	TEDSManIDArray[20].ManID = 22;
	TEDSManIDArray[20].ManName = "Endevco Corporation";
	TEDSManIDArray[21].ManID = 12486;
	TEDSManIDArray[21].ManName = "Trafag AG";
	TEDSManIDArray[22].ManID = 10156;
	TEDSManIDArray[22].ManName = "mg-sensor GmbH";
	TEDSManIDArray[23].ManID = 15095;
	TEDSManIDArray[23].ManName = "Soundtec GmbH";
	TEDSManIDArray[24].ManID = 555;
	TEDSManIDArray[24].ManName = "Druck- und Durchflussmessung";
	TEDSManIDArray[25].ManID = 83;
	TEDSManIDArray[25].ManName = "Dewetron GmbH";
	TEDSManIDArray[26].ManID = 1104;
	TEDSManIDArray[26].ManName = "UNIPULSE Corporation";
	TEDSManIDArray[27].ManID = 1101;
	TEDSManIDArray[27].ManName = "yurbuds Sport Earphones";
	TEDSManIDArray[28].ManID = 1098;
	TEDSManIDArray[28].ManName = "Baker-Calling, Inc.";
	TEDSManIDArray[29].ManID = 107;
	TEDSManIDArray[29].ManName = "Tovey Engineering Inc";
	TEDSManIDArray[30].ManID = 73;
	TEDSManIDArray[30].ManName = "Connection Technology Center";
	TEDSManIDArray[31].ManID = 108;
	TEDSManIDArray[31].ManName = "Ashcroft Inc";
	TEDSManIDArray[32].ManID = 96;
	TEDSManIDArray[32].ManName = "B&W Sensing Technologies";
	TEDSManIDArray[33].ManID = 85;
	TEDSManIDArray[33].ManName = "Adcon Telemetry GmbH";
	TEDSManIDArray[34].ManID = 100;
	TEDSManIDArray[34].ManName = "LLC GlobalTest";
	TEDSManIDArray[35].ManID = 64;
	TEDSManIDArray[35].ManName = "MTS Systems Corp.";
	TEDSManIDArray[36].ManID = 76;
	TEDSManIDArray[36].ManName = "The Pinax Group, Inc.";
	TEDSManIDArray[37].ManID = 77;
	TEDSManIDArray[37].ManName = "Kulite Semiconductor Products Inc";
	TEDSManIDArray[38].ManID = 88;
	TEDSManIDArray[38].ManName = "Streamline Automation, LLC";
	TEDSManIDArray[39].ManID = 31;
	TEDSManIDArray[39].ManName = "HBM";
	TEDSManIDArray[40].ManID = 50;
	TEDSManIDArray[40].ManName = "Waycon";
	TEDSManIDArray[41].ManID = 52;
	TEDSManIDArray[41].ManName = "Pressure Systems";
	TEDSManIDArray[42].ManID = 18;
	TEDSManIDArray[42].ManName = "The Modal Shop, Inc.";
	TEDSManIDArray[43].ManID = 57;
	TEDSManIDArray[43].ManName = "Lorenz Messtechnik GmbH";
	TEDSManIDArray[44].ManID = 23;
	TEDSManIDArray[44].ManName = "PCB Piezotronics, Inc.";
	TEDSManIDArray[45].ManID = 38;
	TEDSManIDArray[45].ManName = "Celesco";
	TEDSManIDArray[46].ManID = 41;
	TEDSManIDArray[46].ManName = "Watlow";
	TEDSManIDArray[47].ManID = 28;
	TEDSManIDArray[47].ManName = "Entek IRD";
	TEDSManIDArray[48].ManID = 32;
	TEDSManIDArray[48].ManName = "TEAC";
	TEDSManIDArray[49].ManID = 17;
	TEDSManIDArray[49].ManName = "Bruel & Kjaer";
	TEDSManIDArray[50].ManID = 1889;
	TEDSManIDArray[50].ManName = "PMT GmbH";
	TEDSManIDArray[51].ManID = 1109;
	TEDSManIDArray[51].ManName = "General Motors";
	TEDSManIDArray[52].ManID = 15913;
	TEDSManIDArray[52].ManName = "VALCOM CO.,LTD.";
	TEDSManIDArray[53].ManID = 12385;
	TEDSManIDArray[53].ManName = "TrigasDM GmbH";
	TEDSManIDArray[54].ManID = 5688;
	TEDSManIDArray[54].ManName = "Duerr Somac GmbH";
	TEDSManIDArray[55].ManID = 3123;
	TEDSManIDArray[55].ManName = "Humanetics Innovative Solutions, Inc";
	TEDSManIDArray[56].ManID = 14908;
	TEDSManIDArray[56].ManName = "Flintec";
	TEDSManIDArray[57].ManID = 1111;
	TEDSManIDArray[57].ManName = "DPA Microphones A/S";
	TEDSManIDArray[58].ManID = 1103;
	TEDSManIDArray[58].ManName = "LMS";
	TEDSManIDArray[59].ManID = 1100;
	TEDSManIDArray[59].ManName = "Nakamura Mfg.Co.,Ltd";
	TEDSManIDArray[60].ManID = 1099;
	TEDSManIDArray[60].ManName = "ME-Messsysteme GmbH";
	TEDSManIDArray[61].ManID = 80;
	TEDSManIDArray[61].ManName = "Microtech Gefell GmbH";
	TEDSManIDArray[62].ManID = 89;
	TEDSManIDArray[62].ManName = "NTI AG";
	TEDSManIDArray[63].ManID = 95;
	TEDSManIDArray[63].ManName = "ACO, Co., Ltd";
	TEDSManIDArray[64].ManID = 82;
	TEDSManIDArray[64].ManName = "Studio Technologies, Inc.";
	TEDSManIDArray[65].ManID = 84;
	TEDSManIDArray[65].ManName = "Natec Sensors GmbH";
	TEDSManIDArray[66].ManID = 98;
	TEDSManIDArray[66].ManName = "Proctor & Chester Measurements Ltd";
	TEDSManIDArray[67].ManID = 69;
	TEDSManIDArray[67].ManName = "Sensy SA";
	TEDSManIDArray[68].ManID = 78;
	TEDSManIDArray[68].ManName = "Showa Measuring Instruments Co., Ltd";
	TEDSManIDArray[69].ManID = 29;
	TEDSManIDArray[69].ManName = "Scitefair International, Inc.";
	TEDSManIDArray[70].ManID = 58;
	TEDSManIDArray[70].ManName = "Kyowa Electronic Instruments Co., Ltd";
	TEDSManIDArray[71].ManID = 27;
	TEDSManIDArray[71].ManName = "Oceana Sensor Technologies";
	TEDSManIDArray[72].ManID = 61;
	TEDSManIDArray[72].ManName = "Metra Mess-und Frequenztechnik Radebeul";
	TEDSManIDArray[73].ManID = 39;
	TEDSManIDArray[73].ManName = "Weed Instrument";
	TEDSManIDArray[74].ManID = 43;
	TEDSManIDArray[74].ManName = "Measurement Specialties";
	TEDSManIDArray[75].ManID = 55;
	TEDSManIDArray[75].ManName = "Futek Advanced Sensor Technology, Inc.";
	TEDSManIDArray[76].ManID = 40;
	TEDSManIDArray[76].ManName = "SensoTec";
	TEDSManIDArray[77].ManID = 25;
	TEDSManIDArray[77].ManName = "Larson Davis";
	TEDSManIDArray[78].ManID = 33;
	TEDSManIDArray[78].ManName = "RION CO.";
	TEDSManIDArray[79].ManID = 3824;
	TEDSManIDArray[79].ManName = "Nanotok LLC";
	TEDSManIDArray[80].ManID = 9914;
	TEDSManIDArray[80].ManName = "MecSense Messtechnik GmbH";
	TEDSManIDArray[81].ManID = 16172;
	TEDSManIDArray[81].ManName = "Unity Surveillance, Inc.";
	TEDSManIDArray[82].ManID = 11699;
	TEDSManIDArray[82].ManName = "Mantracourt Electronics Ltd";
	TEDSManIDArray[83].ManID = 10874;
	TEDSManIDArray[83].ManName = "STS Sensor Technik Sirnach AG";
	TEDSManIDArray[84].ManID = 6644;
	TEDSManIDArray[84].ManName = "Siemens AG, MO MLT BG";
	TEDSManIDArray[85].ManID = 16284;
	TEDSManIDArray[85].ManName = "ZETLAB Company";
	TEDSManIDArray[86].ManID = 1110;
	TEDSManIDArray[86].ManName = "Druck- und Durchflussmessung";
	TEDSManIDArray[87].ManID = 6457;
	TEDSManIDArray[87].ManName = "CHENGDU CAIC ELECTRONICS CO.,LTD";
	TEDSManIDArray[88].ManID = 10700;
	TEDSManIDArray[88].ManName = "Hangzhou Aihua Instruments Co., Ltd.";
	TEDSManIDArray[89].ManID = 1120;
	TEDSManIDArray[89].ManName = "DJB Instruments";
	TEDSManIDArray[90].ManID = 1119;
	TEDSManIDArray[90].ManName = "SIMTEC Elektronik GmbH";
	TEDSManIDArray[91].ManID = 1113;
	TEDSManIDArray[91].ManName = "Valens Semiconductor Ltd.";
	TEDSManIDArray[92].ManID = 103;
	TEDSManIDArray[92].ManName = "YMC Piezotronics, Inc.";
	TEDSManIDArray[93].ManID = 90;
	TEDSManIDArray[93].ManName = "Ridgetop Group Inc.";
	TEDSManIDArray[94].ManID = 92;
	TEDSManIDArray[94].ManName = "A.S.T. Agewandte System Technik GmbH";
	TEDSManIDArray[95].ManID = 97;
	TEDSManIDArray[95].ManName = "ETH-messtechnik gmbh";
	TEDSManIDArray[96].ManID = 81;
	TEDSManIDArray[96].ManName = "Aikoh Engineering";
	TEDSManIDArray[97].ManID = 70;
	TEDSManIDArray[97].ManName = "Druck Ltd.";
	TEDSManIDArray[98].ManID = 75;
	TEDSManIDArray[98].ManName = "Sensordata Technologies, Inc.";
	TEDSManIDArray[99].ManID = 63;
	TEDSManIDArray[99].ManName = "Mettler-Toledo GmbH Process Analytics";
	TEDSManIDArray[100].ManID = 67;
	TEDSManIDArray[100].ManName = "MSC Automotive Ltd";
	TEDSManIDArray[101].ManID = 51;
	TEDSManIDArray[101].ManName = "AVL List GmbH";
	TEDSManIDArray[102].ManID = 19;
	TEDSManIDArray[102].ManName = "Kistler Instrument Corporation";
	TEDSManIDArray[103].ManID = 45;
	TEDSManIDArray[103].ManName = "Telemonitor";
	TEDSManIDArray[104].ManID = 46;
	TEDSManIDArray[104].ManName = "Lion Precision";
	TEDSManIDArray[105].ManID = 53;
	TEDSManIDArray[105].ManName = "ONO SOKKI";
	TEDSManIDArray[106].ManID = 24;
	TEDSManIDArray[106].ManName = "G.R.A.S. Sound & Vibration";
	TEDSManIDArray[107].ManID = 20;
	TEDSManIDArray[107].ManName = "Macro Sensors";
	TEDSManIDArray[108].ManID = 5040;
	TEDSManIDArray[108].ManName = "G-SNS GmbH";
	TEDSManIDArray[109].ManID = 2614;
	TEDSManIDArray[109].ManName = "OOO L Card";
	TEDSManIDArray[110].ManID = 3528;
	TEDSManIDArray[110].ManName = "HAINBUCH GmbH SPANNENDE TECHNIK";
	TEDSManIDArray[111].ManID = 2524;
	TEDSManIDArray[111].ManName = "PR├£FTECHNIK Dieter Busch AG";
	TEDSManIDArray[112].ManID = 5346;
	TEDSManIDArray[112].ManName = "MinebeaMitsumi Inc.";
	TEDSManIDArray[113].ManID = 9588;
	TEDSManIDArray[113].ManName = "Stankowitz Test Equipment GmbH";
	TEDSManIDArray[114].ManID = 1116;
	TEDSManIDArray[114].ManName = "acam-messelectronic gmbh";
	TEDSManIDArray[115].ManID = 1115;
	TEDSManIDArray[115].ManName = "MICRO-EPSILON MESSTECHNIK GmbH & Co. KG";
	TEDSManIDArray[116].ManID = 1112;
	TEDSManIDArray[116].ManName = "Veridt, Inc";
	TEDSManIDArray[117].ManID = 1108;
	TEDSManIDArray[117].ManName = "CSM GmbH";
	TEDSManIDArray[118].ManID = 1106;
	TEDSManIDArray[118].ManName = "Dynamic Systems Inc";
	TEDSManIDArray[119].ManID = 104;
	TEDSManIDArray[119].ManName = "Sentek Dynamics";
	TEDSManIDArray[120].ManID = 101;
	TEDSManIDArray[120].ManName = "L.S. Starrett Company";
	TEDSManIDArray[121].ManID = 91;
	TEDSManIDArray[121].ManName = "Fenda Technology Ltd.";
	TEDSManIDArray[122].ManID = 87;
	TEDSManIDArray[122].ManName = "SVANTEK Sp. z.o.o";
	TEDSManIDArray[123].ManID = 86;
	TEDSManIDArray[123].ManName = "LCM Systems Ltd";
	TEDSManIDArray[124].ManID = 99;
	TEDSManIDArray[124].ManName = "HEAD acoustics GmbH";
	TEDSManIDArray[125].ManID = 62;
	TEDSManIDArray[125].ManName = "Ametek Denmark A/S";
	TEDSManIDArray[126].ManID = 65;
	TEDSManIDArray[126].ManName = "Viatran Corporation";
	TEDSManIDArray[127].ManID = 71;
	TEDSManIDArray[127].ManName = "Tokyo Sokki Kenkyujo Co., Ltd";
	TEDSManIDArray[128].ManID = 74;
	TEDSManIDArray[128].ManName = "BSWA Technology Co., Ltd";
	TEDSManIDArray[129].ManID = 60;
	TEDSManIDArray[129].ManName = "Sherborne Sensors Limited";
	TEDSManIDArray[130].ManID = 35;
	TEDSManIDArray[130].ManName = "Dytran Instruments";
	TEDSManIDArray[131].ManID = 79;
	TEDSManIDArray[131].ManName = "CEA";
	TEDSManIDArray[132].ManID = 21;
	TEDSManIDArray[132].ManName = "Wilcoxon Research";
	TEDSManIDArray[133].ManID = 56;
	TEDSManIDArray[133].ManName = "Knick Elektronische Messgeraete GmbH Co. KG";
	TEDSManIDArray[134].ManID = 30;
	TEDSManIDArray[134].ManName = "National Instruments";
	TEDSManIDArray[135].ManID = 36;
	TEDSManIDArray[135].ManName = "SincoTec";
	TEDSManIDArray[136].ManID = 26;
	TEDSManIDArray[136].ManName = "Key Transducers, Inc.";
	TEDSManIDArray[137].ManID = 48;
	TEDSManIDArray[137].ManName = "RDP Electronics";
	TEDSManIDArray[138].ManID = 15514;
	TEDSManIDArray[138].ManName = "HITEC Sensor Developments";
	TEDSManIDArray[139].ManID = 15548;
	TEDSManIDArray[139].ManName = "WHITEvoid GmbH";
	TEDSManIDArray[140].ManID = 12081;
	TEDSManIDArray[140].ManName = "Fatri-UTC.Inc";
}

src/TEDDY.cpp
/*
 TEDDY.cpp
 
 Transducer Electronic Datasheet Decryption Yielder - An arduino library to read and decode TEDS from acceleration sensors.
 
 */

#include "TEDDY.h"

TEDDY::TEDDY(int busPinInit){
	busPin = busPinInit;
	POPULATETEDDYMANID();
}

TEDDY::~TEDDY(){
   
}

uint8_t TEDDY::read(){
	
	int ReadFailed = 1;
	
	// Read the TEDS ROM to determine the EEPROM device type
	OneWire SensorBus(busPin);
	SensorBus.reset();
	SensorBus.write(0x33); // Read ROM command
	
	for(int i = 0; i < 8; i++) {
		TEDSROM[i] = SensorBus.read();
		if(TEDSROM[i]!=0xFF){
			ReadFailed = 0;
		}
	}
	
	// Assign ROM device type
	ROMDeviceType = TEDSROM[0];
	
	// No TEDS device attached (line was not pulled low -> all bits are 1)
	if(ReadFailed){
		return ReadFailed;
	}
	
	ReadFailed = 1;
	
	// Read the TEDS EEPROM content
	
	// For Kistler 8688A50T (TEDS v.0.9 Legacy - Template Id 0)
	if(ROMDeviceType==0x14 || ROMDeviceType==0x2D){

		SensorBus.reset();
		SensorBus.write(0xCC); // SKIP ROM command
		SensorBus.write(0xF0); // READ MEMORY command
		SensorBus.write(0x00); // STARTING ADDRESS? command
		SensorBus.write(0x00); // STARTING ADDRESS? command
	
		for (int i = 0; i < 32; i++) {
			TEDSData[i] = SensorBus.read();
			if(TEDSData[i]!=0xFF){
				ReadFailed = 0;
			}
		}
	}
	
	// No TEDS device attached (line was not pulled low -> all bits are 1)
	if(ReadFailed){
		return ReadFailed;
	}
	
	// Legacy v.0.9 Template Id 0 - DS2430A (0x14)
	if(ROMDeviceType==0x14){
		offset = 0;
	}
	
	// Legacy v.0.9 Template Id 0 - DS2431 (0x2D)
	if(ROMDeviceType==0x2D){
		offset = 1;
	}
	
	return 0;
}

char* TEDDY::getROMDeviceType(){
	
	if(ROMDeviceType==0x14){
		return "DS2430A (0x14)";
	}
	
	if(ROMDeviceType==0x2D){
		return "DS2431 (0x2D)";
	}
	
	if(ROMDeviceType==0xFF){
		return "TEDS read error (0xFF)";
	}
	
	return "Unknown device";
}

unsigned long int TEDDY::getSerialNumber(){
	
	// Communication error, return 0
	if(ROMDeviceType==0xFF){
		return 0;
	}
	
	return (((TEDSData[7+offset]<<16) + (TEDSData[6+offset]<<8) + TEDSData[5+offset])<<1) + (TEDSData[4+offset]>>7);
}

char TEDDY::getSensorDirection(){
	
	// Communication error, return E
	if(ROMDeviceType==0xFF){
		return 'E';
	}
	
	SensorDirection = ((TEDSData[16+offset]<<7)>>6) + (TEDSData[15+offset]>>7);
	
	if(SensorDirection==0){
		return 'X';
	}
	
	if(SensorDirection==1){
		return 'Y';
	}
	
	if(SensorDirection==2){
		return 'Z';
	}
	
	return 'E';
}

float TEDDY::getSensitivity(){
	
	byte ByteTemp;
	
	// Communication error, return 0
	if(ROMDeviceType==0xFF){
		return 0;
	}
	
	ByteTemp = TEDSData[13+offset]<<6;
	ByteTemp = ByteTemp>>6;
	
	return 0.0001 * pow(1.0002, (ByteTemp<<14) + (TEDSData[12+offset]<<6) + (TEDSData[11+offset]>>2));
}

String TEDDY::getTransducerType(){
	
	// Communication error, return error
	if(ROMDeviceType==0xFF){
		return "TEDS read error (0xFF)";
	}
	
	byte ByteTemp = TEDSData[3+offset]<<4;
	byte MNByte = ByteTemp>>4;
	ByteTemp = TEDSData[4+offset]<<7;
	byte MVByte = ByteTemp>>3;
	byte MVNByte = TEDSData[4+offset]<<1;
	
	ModelNumber = (MNByte<<12) + (TEDSData[2+offset]<<4) + (TEDSData[1+offset]>>4);
	TransducerType = ModelNumber;
	
	ModelVersion = (MVByte + (TEDSData[3+offset]>>4));
	if(ModelVersion>0 && ModelVersion<27){
		ModelVersion += 64; //Offset to C char table
	}
	else{
		return "Error: Unimplemented chr5 value!";
	}
	TransducerType += ModelVersion;
	
	ModelVersionNumber = MVNByte>>2;
	TransducerType += ModelVersionNumber;
	
	return TransducerType;	
}

String TEDDY::getManufacturer(){
	
	// Communication error, return error
	if(ROMDeviceType==0xFF){
		return "TEDS read error (0xFF)";
	}

	byte ByteTemp = TEDSData[1+offset]<<4;
	int ManufacturerID = (ByteTemp<<4) + (TEDSData[0+offset]);
	int i = 0;
	
	while(TEDSManIDArray[i].ManID!=ManufacturerID){
		i++;
	}
	
	return TEDSManIDArray[i].ManName;
}

unsigned int TEDDY::getCalibrationDate(){
	
	// Communication error, return 0
	if(ROMDeviceType==0xFF){
		return 0;
	}
	
	byte ByteTemp = TEDSData[11+offset]<<6;
	
	return (ByteTemp<<10) + (TEDSData[10+offset]<<6) + (TEDSData[9+offset]>>2);
}